### ISSUE TITLE
Object curlies: File/project–wide or standard?

### ISSUE THREAD

**Issue by mightyiam** [standard/eslint-config-standard#35]
So, we got #21 and that's great.

It got us instance–wide consistency.

Now, what about file–wide consistency, as was mentioned [here](https://github.com/feross/standard/issues/182#issuecomment-125738773)?

And, now that I'm thinking about it, why not project–wide consistency?

And thinking further, why not choose a style as standard?

Either no spaces (`{a, b}`) or a single space (`{ a, b }`).

**Comment by timdp** [comment]
Personally, I've been struggling with the choice between `{foo, bar}` and `{ foo, bar }` for a while. I think the former makes more sense because I also write `[foo, bar]`, but it feels like I'm seeing the latter more often. Ultimately, I'm fine with either, but I'm basically expecting Standard to decide for me. Just my two cents.

**Comment by cayuu** [comment]
Agreed. I've seen a lot of inconsistency in a multi-person Standard project with different bikeshed preferences  ensuring _both_ styles `{a}` and `{ a }` are being utilised in different files. When team `{a}` players land on `{ a }` files (and vice versa), I'm seeing commits that do nothing other than change the spacing. 😜 

We've rolled a custom rule to prevent this. But like, so non-Standard.

Also don't care which style prevails, but picking one seems squarely in Standard's decision space.

**Comment by PrototypeAlex** [comment]
I thought I was alone on struggling to pick a side on this, but it seems the lack of clear choice is quite common. 

Is the hard part what @timdp pointed out? 
The fact that choosing a rule for object culries might force a rule for arrays? or can both live in harmony being different from one another?

```javascript
const array = [one, two];
const object = { one, two }; 
```

```javascript
const array = [one, two];
const object = {one, two}; 
```

```javascript
const array = [ one, two ];
const object = { one, two }; 
```

**Comment by dcousens** [comment]
@feross,  its time to pick something for the next version IMHO.

**Comment by jokeyrhyme** [comment]
Apparently the `--fix` option works: http://eslint.org/docs/rules/object-curly-spacing

Surely the fact that compliance can be automated like this makes this a less controversial change? Why would you bring 1 or 2 files into compliance rather than the whole project, when doing the whole project is this simple?

**Comment by feross** [comment]
@dcousens I'm fine with going with your latter example, since that's more consistent. I usually add spaces around my objects, but consistency is more important. If `--fix` can handle it, then it's probably fine to just ship this.

**Comment by timdp** [comment]
Looks like I'm going to be fixing a lot of scripts soon, but all the better! :+1:

**Comment by dcousens** [comment]
@feross my only aversion to the latter is

``` js
let map = {a:1}
```

instead of

``` js
let map = { a: 1 }
```

I think the spacing in my mind maps to a kind of "spatial mapping",  whereas the closely ordered array `[1, 2, 3, 4, 5]` is because the index is implied,  and not specified.

The latter option would also mean:
``` js
let x = {y}
```

This "consistency" might actually lead to mistakes since it looks *closer* to an array.
Maybe.

I leave it to you @feross :+1:

**Comment by feross** [comment]
You raise a good point. This is definitely very subjective. 😅 

Btw, it would be:

```js
let map = {a: 1}
```

Not:

```js
let map = {a:1}
```

**Comment by dcousens** [comment]
I wonder if either is "easier" to search for? 
I have fallen in love with `function myfunc ()` for that reason alone.

**Comment by cayuu** [comment]
`{ a: 1 }` is more readable, `{a: 1}` is more consistent (with other single line enclosure syntax e.g.: `[x, ..]`, `(x, ..)`). This is pure convention, so either choice is "valid". But if consistency is the yardstick, the tight wrap has an edge.

**Comment by dcousens** [comment]
Agreed with @cayuu.
But I'd add a point from my post above.
If safety from errors/mistake is the yardstick.  I think `{ a: 1 }` has the edge.

**Comment by lolmaus** [comment]
This line:

```js
import {belongsTo/*, hasMany*/} from 'ember-data/relationships'
```

triggers this error:

> Expected consistent spacing (standard/object-curly-even-spacing)

I believe it should be considered valid: it contains no stray whitespace; a comment is not whitespace.

**Comment by lolmaus** [comment]
Should I file a separate issue? Into which repo?

**Comment by mightyiam** [comment]
@lolmaus I would file it against [this repo](https://github.com/eslint/eslint/blob/master/lib/rules/object-curly-spacing.js).

**Comment by lolmaus** [comment]
@mightyiam The offending rule is `standard/object-curly-even-spacing`. I believe `object-curly-spacing` does not have that effect.

**Comment by mightyiam** [comment]
Woops

**Comment by feross** [comment]
@lolmaus Yes, this is a separate issue. Please file it here: https://github.com/xjamundx/eslint-plugin-standard

**Comment by lolmaus** [comment]
@feross Yessir! xjamundx/eslint-plugin-standard/issues/13

**Comment by xjamundx** [comment]
@dcousens in your example up there was that about destructuring?
`let { x } = y`

I'd be pretty happy doing file-wide enforcement. I was the one that wrote that original instance-wide enforcement before and probably should have done file-wide instead.

My personal pref is `{ x: 1 }` and `[1, 2]` as well.

**Comment by dcousens** [comment]
@xjamundx no, it was about object definitions in general

**Comment by feross** [comment]
No spaces (e.g. `{x: 1}`, `let {x} = y`, and `[1, 2]`) is more consistent.

But there's an argument to be made that `{ x: 1 }`, `let { x } = y`, and `[1, 2]` make more sense since there's another rule that requires spaces around `{ }` in blocks, e.g. `a => { return a + 1 }`.

And like @dcousens says, it makes objects visually distinctive from arrays.

This is a toss-up. I don't currently feel strongly one way or the other.

**Comment by feross** [comment]
I think that `standard` should aim for consistency whenever possible, so I'm leaning towards `{x: 1}`, `let {x} = y`, and `[1, 2]`.

**Comment by jprichardson** [comment]
My preferred style is `{ x: 1 }`, `let { x } = y`, and `[1, 2]` as well. Since we have spaces for blocks and spaces everywhere else other than `[1, 2]`... maybe it makes sense to `[ 1, 2 ]`? Gah, but then that seems weird with `someFunc(a, b)` and not `someFunc( a, b )`. There's no easy answer here.

**Comment by feross** [comment]
@jprichardson Good points. Seems that no spaces (except for blocks, which are pretty different) is the most consistent, no?

**Comment by timdp** [comment]
I knew I shouldn't have started changing all my code from `{foo}` to `{ foo }`. But fundamentally, I agree: `{foo}` should be treated like `[foo]`. That was also my original rationale.

What made me change my mind aside from this discussion was that objects often add identifiers:

```js
import {foo as f, bar as b} from 'baz'

const [f, b] = baz
const {foo: f, bar: b} = baz

quux(({foo: f, bar: b}) => {
})
```

Even if you ignore the bracket infestation in the last example, they're pretty hard to read without an additional space. Conversely, if you do add the space, you'd have to do it everywhere, like @jprichardson pointed out. That would be especially apparent when destructuring more complex values containing both arrays and objects (which I'm not saying is a good idea).

Bottom line: I'm torn as well but I wanted to weigh in with some additional examples that would be affected.

**Comment by feross** [comment]
@timdp Thanks for the additional examples. I think all but your last example have good readability. I think it will just take some getting used to. (It's not my current style either.)

You're right that the last example isn't ideal. But it's not too much better even *with* the space IMO because there's just a lot going on:

```js
quux(({ foo: f, bar: b }) => {
})
```

Thanks for weighing in! 👍

**Comment by timdp** [comment]
@feross No problem!

The last example is actually a pretty common one in disguise: the AWS SDK, which is pretty popular for obvious reasons, returns objects with PascalCased keys. I typically end up destructuring it and renaming the keys to camelCase. (Typically, there'd be a promise somewhere in the mix rather than a callback but the destructuring remains.) Hence, it's important to get this scenario right.

Not advocating either style here. Just wanted to emphasize the use case's relevance.

At any rate, this time, I won't update my code until Standard 9 is actually out. :stuck_out_tongue:

**Comment by zaynv** [comment]
Just a lone man's opinion, but I kind of think that this syntax:

```js
const obj = { a: 'b' }
```

should be enforced instead of 

```js
const obj = {a: 'b'}
```

This may not really be a good argument for it, but I always associated `standard` with being very "spacey". It seems like the motto is "when in doubt, use a space". Some examples:

**After keywords and before opening brace:**

```js
if (condition) {

}
```

**Before and after operators**
```js
const sum = a + b + c
```

**Before function parameters:**

```js
function foo (bar, baz) {

}
```

**Before and after generator function asterisk:**

```js
function * foo (bar, baz) {

}
```

although im not sure how people would feel about enforcing spaces in array declarations like:

```js
const arr = [ 1, 2, 3 ]
```

But yeah at the end of the day, I guess the main thing people care about is just picking something.

**Comment by dcousens** [comment]
``` js
var f = {x: 1, y: 3}
var b = [1, 3, 4]
var u = {
  x: 1,
  y: 3
}

quux(({foo: f, bar: b}) => {
```
or
``` js
var f = { x: 1, y: 3 }
var b = [1, 3, 4]
var u = {
  x: 1,
  y: 3
}
quux(({ foo: f, bar: b }) => {
```

The second case seems less cluttered, and I can readily distinguish the syntax.
The first case just seems unnecessarily "compact".

The declarations of `f` and `u` in the first case also seem inconsistent... somehow.

**Comment by dcousens** [comment]
I think it's also because every instance of `{` usually has white/empty space after it.
~~This would be the first exception to that rule.~~
Ignoring *empty* definitions.

See below:

``` json
var a = {}
var b = {a: 1}
var bb = {a}
var c = {
  a: 1
}
var d = function () {}
var e = function () {

}
var f = function ({a}) {

}
var g = function ({a, b, c: 1}) {

}
```

vs

``` json
var a = {}
var b = { a: 1 }
var bb = { a }
var c = {
  a: 1
}
var d = function () {}
var e = function () {

}
var f = function ({ a }) {

}
var g = function ({ a, b, c: 1 }) {

}
```

**Comment by jokeyrhyme** [comment]
Within my own code, I'm very much in the curly-space camp, for sure. I agree with @zaynv 's argument about this already fitting Standard's conventions.

I'm also theoretically in favour of spacey-arrays, too, for consistency with curly-spaces. But, in practice, my own code shows a 50-50 split of spacey-arrays and spaceless-arrays. /shrug

**Comment by wuhkuh** [comment]
Exactly, using spaces around the curly braces would make sense, as this makes complex items more readable and the current style of standard feels spacey.

Besides, we should let the build process minify and uglify the arrays and objects, and keep our focus on making things **readable** for ourselves. That is the entire thing we should be focused on - keeping things readable, even for people who are new to the source code of our projects. Why else would we implement a standard?

My vote goes to both spacey-arrays and spacey-objects.

**Comment by timdp** [comment]
@wuhkuh Readable *and consistent*. The point here is that some expressions have spaces and some don't, and whichever choice gets made here in the end is going to violate some of those.

**Comment by wuhkuh** [comment]
@timdp I think spacey-arrays and spacey-objects would be a perfect solution to this question. As for the function arguments:

`function write (callback)` would still be valid. We could say that abstractions do not need spaces.

When you call the function `write({ "hey": 1 })` the input would not be an abstraction anymore as the input is atomic.

So what if we input `write(object)`? Well that follows the rule of an abstraction, and therefore does not need a space.

Please provide me with feedback on this idea.

**Comment by timdp** [comment]
That would imply spaces inside square braces too, among other things.

**Comment by wuhkuh** [comment]
@timdp That's exactly what I'm suggesting.

`let arr = [ 1, 2, 3 ]`
`let obj = { 1, arr, 3, 2 }`

`function write (item)`
`write(obj)`

The difference is the abstraction 'item'. I can't find the official term for this, but I'm sure you understand my intentions.

**Comment by dcousens** [comment]
If we went with spacey-arrays, my brain is going to haemorrhage for a while when I see `[ x ]` in the wild.
Since ES6, I've indoctrinated myself to see `{ x }` as equivalent to `{ x: x }`.

**Comment by jokeyrhyme** [comment]
Do we think there is sufficient consensus to move forward with spacey-objects?
If so, should we PR this baby and move the array discussion over to a new Issue?

**Comment by feross** [comment]
I want to hold off on any changes for a while. We just released a major version and I'm concerned about users getting change fatigue.

My take: Let's all go spend our time building things. We can revisit all this in early 2017 and figure what new rules we want to add in `standard` v9.

**Comment by timdp** [comment]
You could already decide and commit to whatever choice you make for v9 though. Knowing about this issue puts me in doubt whenever I type a curly brace, which isn't great either. :-)

**Comment by jokeyrhyme** [comment]
Yeah, we could use GitHub Issue Milestones to track things that are probably going to be in 9.0

**Comment by dcousens** [comment]
@feross indeed, even if the change isn't finalised until `v9`,  that ambiguity is a bit lame.

**Comment by feross** [comment]
The thing is that I want to look at what percentage of current standard users use each style, just so we don't pick the option that only 20% of users are using, or something like that. That's going to take a bit of time to look into.

Sorry for the current ambiguity. If it's any consolation, this rule is automatically fixable with `--fix`, so it'll be easy to switch!

**Comment by cayuu** [comment]
> just so we don't pick the option that only 20% of users are using

Yes, this. While we'd make the internal switch to an unusual pattern (spaced arrays), and I _sympathise_ with the desire to factor "readability" into decisions, using "readability" (phrased well as being "spacey") as the primary yardstick becomes a valid argument for pushing other uncommon-in-JS patterns on this basis (eg. open blocks on new lines like in C). The "uncommon-in-JS" is my pause.

Consistency with _common practice_ matters, regardless if it is subjectively/conditionally "worse" (hello bikeshed) - don't surprise people. That said, if 20%+ of JS projects use padded arrays, and there are some reference "big projects" that use this style (ie. momentum towards this), then it's less of an issue (as this reduces the expected-cost of the inevitable delays/screwups/confusion/wat due to unfamiliarity).

FWIW the internal style we ended up on was the "tight" wrap `{x, y, z}` for consistency with other enclosure syntax, and a _caveat_ that if it got too crazy, then instead of padding the enclosure, the elements got newlined (which matches how we've handled array formatting since the stone age).

**Comment by jokeyrhyme** [comment]
I wonder if there's a query we can run that will tell us what the common style choice is?
https://cloud.google.com/bigquery/public-data/github

**Comment by timdp** [comment]
@cayuu Great point about the newlines. That's basically what I've been doing.

**Comment by kilianc** [comment]
My two cents:

* `{ a: 'foo' }`
* `[1, 2, 3]`

But I would be ok with

* `{ a: 'foo' }`
* `[ 1, 2, 3 ]`

for consistency.

`{a: 'foo'}` makes it harder for me to scan things and it gets messier if it's just `{a}` :(

**Comment by thejohnfreeman** [comment]
I see this issue hasn't had any discussion in almost 4 months. Regarding @feross's [comment](https://github.com/feross/eslint-config-standard/issues/35#issuecomment-248507484), "We can revisit all this in early 2017", maybe now is the time.

After reading all the discussion, my preference has converted from spaceless-objects to spacey-objects, but now I also want spacey-arrays for a few reasons already mentioned:

1. Consistency with spacey-objects.
2. Fits the tradition and convention of `standard`. @zaynv [said](https://github.com/feross/eslint-config-standard/issues/35#issuecomment-246897946) "I always associated standard with being very 'spacey'."
3. Consistency with arrays split over multiple lines. @dcousens [said](https://github.com/feross/eslint-config-standard/issues/35#issuecomment-246901984):

```js
var u = {
  x: 1,
  y: 3
}
var f = {x: 1, y: 3}
```

> The declarations of f and u in the first case also seem inconsistent... somehow.

Well, we can do similar with arrays:

```js
var a = [
  x,
  y
]
var b = [a, b]
```

Are these similarly inconsistent? Probably. This argument isn't that convincing to me, but it goes with the other two.

**Comment by jokeyrhyme** [comment]
It'd be handy for some accord to be reached between https://github.com/jlongster/prettier and JavaScript Standard

**Comment by mightyiam** [comment]
@jokeyrhyme I would discuss that in another issue.

I'm :+1: on spacey. And on determining this.

**Comment by saadq** [comment]
I'm also on the spacey side. Arrays and object literals are two different things, so I think the "inconsistency" between `[x, y]` and `{ x, y }` makes sense.

**Comment by tunnckoCore** [comment]
What I was thinking about last few months. I'm pretty used to spacey objects, but not spacey arrays (maybe sometimes, hm).

```js
var a = { x: 2 }
var a = {
  x: 1
}
var a = {
  x: 1,
  y: 2
}
var a = [
  x,
  y
]
var b = [x]
var b = [123]
var b = [ x, y ]
var b = [ { a: 'b' }, 2, { c: 3 }, 4 ]
var b = [{ a: 'b' }]
var b = [ { a: 'b' } ] // too much spaces? above is better
var b = {
  a: 1,
  b: {
    c: [ { cc: 22 }, 333 ]
  },
  d: [ 'e', 444, [ 1, 2, 3 ] ]
}
```

If we apply consistency, it may seem too spacey in some times, but in most cases it would be separated in separate lines, so won't maybe won't look too spacey. As about the `array-bracket-spacing` i think we should use [`singleValue: false` option](http://eslint.org/docs/rules/array-bracket-spacing#singlevalue)

**Comment by cayuu** [comment]
It's worth noting that this change (no matter which way we go - but especially spaced-arrays), is likely to affect a significant portion of `standard` users. 

We have three options for concluding this:

1. Defer to @feross for a decision
2. Use the most voted/weighted preference of the 16 current participants
3. Defer to conventions used in popular of ES6+ projects (perhaps looking at the OSS projects from the big companies listed in the README)

My preference would be for 1 or 3. I'm wary of a "happened to be in the room" policy approach (2).

**Comment by dcousens** [comment]
What is the ecosystem impact of either?

**Comment by cayuu** [comment]
Started at the top of the README list of companies, left to right:

- npm: spaced objects, tight arrays
- opbeat: spaced objects, tight arrays

..then I got tired looking ;)

**Comment by jokeyrhyme** [comment]
For some perspective, it looks like the prettier folks couldn't decide and just made this a "bracketSpacing" option:

https://github.com/prettier/prettier#api

**Comment by wuhkuh** [comment]
@jokeyrhyme
Doesn't that defeat the purpose of a standard?
The decisions should be up to this package.

**Comment by jokeyrhyme** [comment]
@wuhkuh I just mention prettier, because it's another example of an opinionated style guide, and it's been a difficult enough decision to make that they've made it an option

In principle, I agree that eliminating pointless bike-shedding and decision-making is the noble goal of an opinionated project

**Comment by wuhkuh** [comment]
Despite the fact that npm and opbeat use tight arrays, I still prefer spacey for the reasons @thejohnfreeman put down. 
It would feel unnatural if you keep the arrays tight while you space out objects and functions with their arguments. When in doubt, space out.

There are also tools available that update existing code to suit the standard. This debunks 'patching is a lot of work!'

I would like to add that if we decide which way to go, we could start by sending warnings instead of erroring when this patch is just introduced.
This would decrease the shock of this big decision at the packages that use this standard.

**Comment by xjamundx** [comment]
paypal uses tight arrays FWIW

**Comment by LinusU** [comment]
Just realized that having spacey arrays is a way to distinguish array literals and property access 🤔 

```js
const array = [ 1 ]
const number = array[0]
```

**Comment by tunnckoCore** [comment]
@LinusU good point. :)

@wuhkuh `console.log(obj[ 'is' ])` is another rule that i think it should enforce non-spacey style - currently is `either` set by eslint-plugin-standard.

**Comment by falmar** [comment]
I see more often about spacey objects than spacey arrays, space arrays feel a bit awkward to me

```js
const obj = { a: 1, b: 2 }
const arr = [1, 2]
```

I personally find this more readable
```js
import React, { Component } from 'react'
import { connect } from 'react-redux'

import { getItems } from '../actions'

const objItems = [{ one: 1 }, { two: 2 }]

const App = () => (
  <div
    style={{ height: '100vh' }}
    objItems={objItems}
    arrItems={[1, 2]}
  />
)

const mapStateToProps = ({ aReducer, bReducer }) => ({ aReducer, bReducer })

export default connect(mapStateToProps, { getItems })(App)
```

Than this
```js
import React, {Component} from 'react'
import {connect} from 'react-redux'

import {getItems} from '../actions'

const objItems = [{one: 1}, {two: 2}]

const App = () => (
  <div
    style={{height: '100vh'}}
    objItems={objItems}
    arrItems={[1, 2]}
  />
)

const mapStateToProps = ({aReducer, bReducer}) => ({aReducer, bReducer})

export default connect(mapStateToProps, {getItems})(App)
```

PD: Sorry for using long noisy examples

**Comment by juliaogris** [comment]
For those of you looking for a quick fix in the meantime:

* `npm install --global eslint eslint-config-standard eslint-config-standard-jsx eslint-plugin-promise eslint-plugin-react eslint-plugin-standard`
* `echo '{"extends": ["standard", "standard-jsx"], "rules": {"object-curly-spacing": [1, "always"]}}' > ~/.eslintrc`

__Careful__, this will replace an exisiting `~/.eslintrc `.
Replace `"always"` with `"never"` if preferred or read up on the ESLint rule [object-curly-spacing](http://eslint.org/docs/rules/object-curly-spacing) and  [array-bracket-spacing](http://eslint.org/docs/rules/array-bracket-spacing).
Then use with:
* `eslint --fix FILE`
*  [ESLint Formatter](https://github.com/TheSavior/ESLint-Formatter) for Sublime

This workaround got rid of the majority of my non-fixable standard.js errors. Apparently, at least for me, inconsistent spacing happens a lot less often with [array-bracket-spacing](http://eslint.org/docs/rules/array-bracket-spacing). So my suggestion is to just define `object-curly-spacing` for the time being and leave array spacing as is (consistent). 

Count me in on spacey objects.

❤️ Standard.js

**Comment by wuhkuh** [comment]
I might have changed my mind on this. An array looks clean already without spaced-style, and spaced arrays make statements and declarations harder to read in some cases.

```js
const Obj = {
  properties: [ { type: Object } ] // this sucks. Lets add an exception
}
const Arr = [ ] // sure
console.log(Obj[ 'properties' ]) // this sucks. Lets add an exception
```

Before you know it, we forgot what we tried to achieve: consistency. If we want to patch those situations we would need exceptions to keep it readable (and convenient), therefore making it less consistent.
For anyone in doubt: enforce spaced arrays for a while. You won't quite enjoy them as much as you'd think. I've done this and it changed my mind completely.

My vote is for spaced objects and tight arrays. Arrays are readable and consistent as they are.
I think the most of us agree on spaced objects as far as this thread goes.

@feross Could you update us on the status?

**Comment by tunnckoCore** [comment]
> console.log(Obj[ 'properties' ]) // this sucks. Lets add an exception

That's totally different rule - `computed-property-spacing`. And is is set to `either` by `eslint-plugin-standard` which means it is allowed, but yea it really sucks that's why I force it to `["error", "never"]`

**Comment by mightyiam** [comment]
I was just mentoring a person. On a project where we use standard. And I chose to just not tell him to space out the objects. It came out like this: `{foo: { bar: 'bar' }}`. So the outer object is tight and the inner is spaced. But I didn't tell him to change it. Because we have a linter. And I am not it. Can we go the route of spaced objects and get on with it?

**Comment by rstacruz** [comment]
I like this a lot. I'm sure it will impact a lot, so maybe for the next major version.

**Comment by alejandrodnm** [comment]
Hey guys it's been almost a year since this issue was opened. I'm going to add a quote from the FAQ

> At the end of the day you have to 'just pick something', and that's the whole philosophy of standard

I think in all this time people have been given the opportunity to express their opinion, but it's time to decided. 

I really don't have a preference, I just want to see this issue close, I have colleges using both approach and I can't enforce a style on them, because as @mightyiam just said, we have a linter for that (That we really like) and the CI passes.

You are always going to find people who prefer one way or the other, but I don't think they will just remove standard from their projects just because they disagree with a rule.

Cheers.

**Comment by feross** [comment]
Rest assured that this will be decided in a future version. There's no need for folks to continue weighing in.

The past few releases have had some pretty large breaking changes, and there's a limit to how many breaking changes we can include in a given release.

I'll lock this issue to save everyone some time.

**Comment by feross** [comment]
Thank you to everyone for sharing your thoughts on this issue. Thanks to @LinusU for the analysis in this comment and for sending a PR: https://github.com/standard/standard/issues/609#issuecomment-400826566

This rule will be included in the `standard` 12 release which will happen later tonight. Happy to merge this and put this one behind us! 🙌
